modules/util.py

```diff
@@ -350,3 +352,138 @@ def mcopy(dev, src, dst):
         log.debug("Failed to mcopy in " + dev);
         return False
     return True
+
+def safeCopy(src, dst, buffer_size=1024*1024):
+    if os.path.isfile(src):
+        return safeFileCopy(src, dst, buffer_size)
+    elif os.path.isdir(src):
+        return safeDirCopy(src, dst, buffer_size)
+    else:
+        log.error("Unknown src target to copy " + src + ".")
+        return False
+
+def safeDirCopy(src, dst, buffer_size=1024*1024):
+    # src must be a dir
+    if not os.path.isdir(src):
+        log.error("Can't copy source as " + src + " is not a directory.")
+        return False
+
+    # dst must not exist
+    if os.path.isdir(dst):
+        log.error("Can't copy to an existent directory " + dst + " .")
+        return False
+
+    # Copy each file in the structure of src to dst
+    for root, dirs, files in os.walk(src):
+            for name in files:
+                srcfullpath = os.path.join(root, name)
+                dstfullpath = os.path.join(dst, os.path.relpath(srcfullpath, src))
+                print(srcfullpath + " -> " + dstfullpath)
+                if not safeFileCopy(srcfullpath, dstfullpath, buffer_size):
+                    return False
+    return True
+
+def safeFileCopy(src, dst, buffer_size=1024*1024):
+    # src must be a file
+    if not os.path.isfile(src):
+        log.error("Can't copy source as " + src + " is not a file.")
+        return False
+
+    # Make sure dst is either non-existent or a file (which we overwrite)
+    if os.path.exists(dst):
+        if os.path.isfile(dst):
+            log.warning("Destination file " + dst + " already exists. Will overwrite.")
+        elif os.path.isdir(dst):
+            log.error("Destination target " + dst + " is a directory.")
+            return False
+        else:
+            log.error("Destination target " + dst + " is unknown.")
+            return False
+
+    # Copy file to src.tmp
+    srcfd = open(src, 'rb')
+    srctmpfd = open(src + ".tmp", "wb")
```

Comment:
The temporary file should be created at the destination, not as a sibling to the source. Otherwise the `rename()` won't work because you'll be crossing a filesystem boundary